addons/marketing_campaign/models/marketing_campaign_workitem.py

```diff
@@ -0,0 +1,227 @@
+                date = fields.Datetime.from_string(fields.Datetime.now())
+
+                for transition in activity.to_ids:
+                    if transition.trigger == 'cosmetic':
+                        continue
+                    launch_date = False
+                    if transition.trigger == 'auto':
+                        launch_date = date
+                    elif transition.trigger == 'time':
+                        launch_date = date + transition._delta()
+
+                    if launch_date:
+                        launch_date = fields.Datetime.to_string(launch_date)
+                    values = {
+                        'date': launch_date,
+                        'segment_id': self.segment_id.id,
+                        'activity_id': transition.activity_to_id.id,
+                        'partner_id': self.partner_id.id,
+                        'res_id': self.res_id,
+                        'state': 'todo',
+                    }
+                    new_wi = self.create(values)
+
+                    # Now, depending on the trigger and the campaign mode
+                    # we know whether we must run the newly created workitem.
+                    #
+                    # rows = transition trigger \ colums = campaign mode
+                    #
+                    #           test    test_realtime     manual      normal (active)
+                    # time       Y            N             N           N
+                    # cosmetic   N            N             N           N
+                    # auto       Y            Y             N           Y
+                    #
+
+                    run = (transition.trigger == 'auto' \
+                            and campaign_mode != 'manual') \
+                          or (transition.trigger == 'time' \
+                              and campaign_mode == 'test')
+                    if run:
+                        new_wi._process_one()
+        except Exception:
+            tb = "".join(format_exception(*exc_info()))
+            self.write({'state': 'exception', 'error_msg': tb})
+
+    @api.multi
+    def process(self):
+        for wi in self:
+            wi._process_one()
+        return True
+
+    @api.model
+    def process_all(self, campaign=None):
+        if campaign is None:
+            campaign = self.env['marketing.campaign'].search([('state', '=', 'running')])
+        for camp in campaign.filtered(lambda rec: rec.mode != 'manual'):
+            while True:
+                domain = [('campaign_id', '=', camp.id), ('state', '=', 'todo'), ('date', '!=', False)]
+                if camp.mode in ('test_realtime', 'active'):
+                    domain += [('date', '<=', fields.Date.today())]
+                    workitem = self.search(domain)
+                    if not workitem.ids:
+                        break
+                    workitem.process()
+        return True
+
+    @api.multi
+    def preview(self):
```

Comment:
If I agree with you that there can be multiple record then what should be return values of this method, is it possible to return multiple actions ? BTW this method is called from form button and it will generally display preview of single record(and you can not return multiple actions :))